### DISCUSSION TITLE
What to do after the poisoning attack?

### DISCUSSION THREAD

**Discussion by schroe-ja** [Trusted-AI/adversarial-robustness-toolbox#1636] [Q&A] [answered]
Hi,

I try to figure out how to use the `PoisoningAttackBackdoor`, `HiddenTriggerBackdoor`, and `PoisoningAttackCleanLabelBackdoor`.
After executing the `poison()` function what do I need to do? I can't really figure that out from the examples.

If I executed the function do I have to put my poisoned examples back into the original training data or do I need to declare an extra poisoned dataset? 

And after testing I have to do a predicition with the poisoned sample right? - Because I can only trigger my backdoor if the pattern is inside an image.

**Comment by beat-buesser** [comment, marked as answer]
Hi @EvilWatermelon

Yes, these poisoning attacks create and return a poisoned dataset or samples by calling the method `poison`.

`PoisoningAttackBackdoor` applies the poison to all samples and you have to mix them with benign samples yourself if you don't want to have all samples poisoned or use in combination with `HiddenTriggerBackdoor` and `PoisoningAttackCleanLabelBackdoor` which have arguments defining the fraction of samples that should be poisonous.

After generating the poisoned dataset you have to used it to train a model. Then evaluate the model on test datasets samples that are benign and on samples containing the backdoor.

This notebook contains the complete workflow: https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/notebooks/poisoning_attack_clean_label_backdoor.ipynb

**Reply by schroe-ja** [reply]
After calling the `poison` function I deleted the original images from the training dataset and then used `np.concatenate()` to add the poisoned samples. If I assign the poisoned dataset as the "original" dataset `X_train = poisoned_x` the training time increase from about 3 - 5 mins to > 10 mins. Do I have to make an additional `model.fit()` like your example showed or is that normal, that my training time increases when I mix the original and the poisoned training data?

**Reply by beat-buesser** [reply]
@EvilWatermelon Theoretically it should not increase the training time, but there might be reasons for it. What is the shape for `X_train` before and after the increase?

Can you share a example script that shows the increase in training time?

**Reply by schroe-ja** [reply]
The shape is `(27446, 30, 30, 3)` before  poisoning. Then I take a random selection of images, for example 100. I delete those pictures from `X_train` and have `(27346, 30, 30, 3)`.

```
n_train = np.shape(x)[0]
num_selection = num_of_images
random_selection=np.random.choice(n_train, num_selection)
temp_x = x[random_selection]

del_x = np.delete(x, [random_selection], axis=0)

y = y[random_selection]

backdoor_class = PoisoningAttackBackdoor(poison_func)   

poisoned_x, poisoned_y = backdoor_class.poison(temp_x, y)

poisoned_data = np.concatenate((del_x, poisoned_x), axis=0)

```

The poisoned ndarray `poisoned_x` and the `del_x` are concatenated into a new ndarray `poisoned_data` with shape `(27446, 30, 30, 3)`. And when I use this new ndarray the training time increases.

**Reply by schroe-ja** [reply]
I changed my ML model to this example:  https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/notebooks/poisoning_attack_clean_label_backdoor.ipynb 

On the proxy.fit() function I got the following error `IndexError: index 11 is out of bounds for axis 1 with size 10` at 
```
\art\utils.py", line 628, in to_categorical
categorical[np.arange(labels.shape[0]), np.squeeze(labels)] = 1
```
I have 43 labels. Do I have to change this anywhere in the example notebook?

**Reply by beat-buesser** [reply]
@EvilWatermelon Have you already updated the number of classes in cell 7: `targets = to_categorical([9], 10)[0]` (10 -> 43)?

**Reply by schroe-ja** [reply]
@beat-buesser I finally fixed it last weekend. I copied your code from the notebook and had to install `CUDA 11.5` and `cuDNN 8.2`.
I needed to change the output layer to 43 and as you said in your last message I needed to change the targets from 10 to 43. 


Here some feedback: I tried to implement these attacks into a SVM with sklearn. I had all of the issues above and it would be a big help for if you explain in the notebooks or in the docs (which are not working) how to implement the backdoor attacks for different ML models if that is possible. If not it would be cool to say that in a short sentence in the docs and/or notebooks.

**Reply by beat-buesser** [reply]
Hi @EvilWatermelon Thank you very much for your feedback! Do you have any specific suggestion how to improve the notebook that you could contribute as part of a pull request?

Could you please share a link to where the docs are not working?

**Reply by schroe-ja** [reply]
Hi @beat-buesser, yes I can improve it and do a pull request in the next few days.

Could you please share a link to where the docs are not working?
https://adversarial-robustness-toolbox.readthedocs.io/en/latest/modules/attacks/poisoning.html

**Reply by beat-buesser** [reply]
Hi @EvilWatermelon Thank you, that would be really great! I'll take a look at the docs build at the link.
